src/xsdata.cpp

```diff
@@ -0,0 +1,700 @@
+#include "xsdata.h"
+
+namespace openmc {
+
+//==============================================================================
+// XsData class methods
+//==============================================================================
+
+XsData::XsData(int energy_groups, int num_delayed_groups, bool fissionable,
+     int scatter_format, int n_pol, int n_azi)
+{
+  int n_ang = n_pol * n_azi;
+
+  // check to make sure scatter format is OK before we allocate
+  if (scatter_format != ANGLE_HISTOGRAM && scatter_format != ANGLE_TABULAR &&
+      scatter_format != ANGLE_LEGENDRE) {
+    fatal_error("Invalid scatter_format!");
+  }
+  // allocate all [temperature][phi][theta][in group] quantities
+  total = double_2dvec(n_ang, double_1dvec(energy_groups, 0.));
+  absorption = double_2dvec(n_ang, double_1dvec(energy_groups, 0.));
+  inverse_velocity = double_2dvec(n_ang, double_1dvec(energy_groups, 0.));
+  if (fissionable) {
+    fission = double_2dvec(n_ang, double_1dvec(energy_groups, 0.));
+    nu_fission = double_2dvec(n_ang, double_1dvec(energy_groups, 0.));
+    prompt_nu_fission = double_2dvec(n_ang, double_1dvec(energy_groups, 0.));
+    kappa_fission = double_2dvec(n_ang, double_1dvec(energy_groups, 0.));
+  }
+
+  // allocate decay_rate; [temperature][phi][theta][delayed group]
+  decay_rate = double_2dvec(n_ang, double_1dvec(num_delayed_groups, 0.));
+
+  if (fissionable) {
+    // allocate delayed_nu_fission; [temperature][phi][theta][in group][delay group]
+    delayed_nu_fission = double_3dvec(n_ang, double_2dvec(energy_groups,
+         double_1dvec(num_delayed_groups, 0.)));
+
+    // chi_prompt; [temperature][phi][theta][in group][delayed group]
+    chi_prompt = double_3dvec(n_ang, double_2dvec(energy_groups,
+         double_1dvec(energy_groups, 0.)));
+
+    // chi_delayed; [temperature][phi][theta][in group][out group][delay group]
+    chi_delayed = double_4dvec(n_ang, double_3dvec(energy_groups,
+         double_2dvec(energy_groups, double_1dvec(num_delayed_groups, 0.))));
+  }
+
+  scatter.resize(n_ang);
+  for (int a = 0; a < n_ang; a++) {
+    if (scatter_format == ANGLE_HISTOGRAM) {
+      scatter[a] = new ScattDataHistogram;
```

Comment:
Since this type is holding a vector of pointers (that it owns), it should have a destructor that deletes them. Alternatively, you could make it a vector of `unique_ptr`s so that they are automatically deleted when out of scope.

Reply:
- [x]